hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/LegacyArchivedMetaEntryReader.java

```diff
@@ -0,0 +1,258 @@
+                    // Filter blocks in desired time window
+                    .map(r -> (GenericRecord) r.getData())
+                    .filter(commitsFilter::apply)
+                    .map(r -> readCommit(r, loadInstantDetails))
+                    .filter(c -> filter == null || filter.isInRange(c.getTimestamp()))
+                    .forEach(instantsInRange::add);
+              }
+            }
+          }
+
+          if (filter != null) {
+            int instantsInCurrentFile = instantsInRange.size() - instantsInPreviousFile;
+            if (instantsInPreviousFile > 0 && instantsInCurrentFile == 0) {
+              // Note that this is an optimization to skip reading unnecessary archived files
+              // This signals we crossed lower bound of desired time window.
+              break;
+            }
+          }
+        } catch (Exception oriException) {
+          // merge small archive files may leave uncompleted archive file which will cause exception.
+          // need to ignore this kind of exception here.
+          try {
+            if (isCorruptCompactionFile(metaClient, fs)) {
+              continue;
+            }
+            throw oriException;
+          } catch (Exception e) {
+            // If anything wrong during parsing merge archive plan, we need to throw the original exception.
+            // For example corrupted archive file and corrupted plan are both existed.
+            throw oriException;
+          }
+        }
+      }
+
+      ArrayList<HoodieInstant> result = new ArrayList<>(instantsInRange);
+      Collections.sort(result);
+      return result;
+    } catch (IOException e) {
+      throw new HoodieIOException(
+          "Could not load archived commit timeline from path " + metaClient.getArchivePath(), e);
+    }
+  }
+
+  public static boolean isCorruptCompactionFile(HoodieTableMetaClient metaClient, FileStatus fileStatus) throws IOException {
+    Path planPath = new Path(metaClient.getArchivePath(), MERGE_ARCHIVE_PLAN_NAME);
+    HoodieWrapperFileSystem fileSystem = metaClient.getFs();
+    if (fileSystem.exists(planPath)) {
+      HoodieMergeArchiveFilePlan plan = TimelineMetadataUtils.deserializeAvroMetadata(FileIOUtils.readDataFromPath(fileSystem, planPath).get(), HoodieMergeArchiveFilePlan.class);
+      String mergedArchiveFileName = plan.getMergedArchiveFileName();
+      if (!StringUtils.isNullOrEmpty(mergedArchiveFileName) && fileStatus.getPath().getName().equalsIgnoreCase(mergedArchiveFileName)) {
+        LOG.warn("Catch exception because of reading uncompleted merging archive file " + mergedArchiveFileName + ". Ignore it here.");
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /**
+   * Sort files by reverse order of version suffix in file name.
+   */
+  public static class ArchiveLogVersionComparator implements Comparator<FileStatus>, Serializable {
```

Comment:
tests

Reply:
It's copied from existing code, should be tested altogether in `TestLegacyArchivedMetaEntryReader`.